WFE: Fix revocation authentication.

cpu opened the pull request:
This commit brings over more of Boulder's WFE2 revocation handling to
Pebble and supports **both** authenticating revocation requests with:
* a JWS with a KeyID specifing the account that previously issued the certificate
* a JWS carrying an embedded JWK containing the certificate to be revoked's public key.

Previously the revocation implementation specified it only
supported authenticating the request with the certificate's key and an
embedded JWK. In practice the revocation logic was still expecting a key
ID and authentication from the ACME account. It also did not verify that 
the certificate to be revoked was originally issued by the account that
authenticates the revocation request.


Resolves https://github.com/letsencrypt/pebble/issues/134

jsha commented:
Since @calavera submitted #130, which added the account-based authentication for revocation, I assume their automated tests probably rely on that account-based authentication. So I think removing it would probably break their tests. Do you agree? If so, I think we should consider that a blocker on this PR, and only land this if we can also preserve account-based authentication.

cpu commented:
>  Do you agree? If so, I think we should consider that a blocker on this PR, and only land this if we can also preserve account-based authentication.

I agree, good point. I'll close this PR and resubmit with account based revocation implemented as well. :+1:

felixfontein commented:
Couldn't most parts of this PR be recycled to implement both checks? I.e. do something along
``` .go
	body, _, prob := wfe.verifyPOST(ctx, logEvent, request, wfe.lookupJWK)
	using_account_key = true;
	if prob != nil {
		// We use extractJWK rather than lookupJWK here because the request is
		// authenticated with an embedded JWK not associated with an account/kid.
		body, key, prob := wfe.verifyPOST(ctx, logEvent, request, wfe.extractJWK)
		if prob != nil {
			wfe.sendError(prob, response)
			return
		}
		using_account_key = false;
	}

	[...]
	
	// Check that the JWK that authenticated the request is equal to the
	// certificate's public key
	if !using_account_key && !keyDigestEquals(key, cert.Cert.PublicKey) {
		wfe.sendError(acme.UnauthorizedProblem(
			"JWK embedded in revocation request must be the same public key as the cert to be revoked"),
			response)
		return
	}
```
in `wfe/wfe.go`? I don't really know the code well enough (and also am not too familiar with Go) and don't have time to try this out right now, so take this with a grain of salt ;-)

cpu commented:
@felixfontein That approach is ~= what Boulder does. I have a branch mostly finished that ports the logic back to Pebble. I need to find time to test it and then I will submit a PR. Hopefully by ~EOD today.

cpu commented:
@jsha @calavera @felixfontein I think this PR is ready for :mag:'s again if you folks have time. I fixed the original account based revocation and added the certificate based revocation. It could certainly benefit from some real-world ACME client testing.

jsha commented on wfe/wfe.go:
```diff
@@ -1615,48 +1661,176 @@ func (wfe *WebFrontEndImpl) RevokeCert(
+		}
+		return acme.UnauthorizedProblem(
+			fmt.Sprintf(
+				"Account %q did not request issuance of the to-be-revoked certificate",
```
"The certificate being revoked is not associated with account %q". The "did not request issuance" is confusing because it sounds like Pebble is complaining that the revocation request isn't an issuance request.

cpu replied:
WFM: Fixed in 6fe0ac0

felixfontein commented:
Thanks for implementing this! I can say that at least my client now successfully revoked a certificate issued by Pebble using its private key. (So far I tested that only against Let's Encrypt staging, and there it worked as well.)
